### PR TITLE
e2e: revisit the e2e test suite

### PR DESCRIPTION
**Pull request by pires**
TL;DR: install `minikube` and run `rm -rf vendor; make dep; make e2e`.

Now, the longer version...

This PR revisits the end-to-end (e2e) test suite in order to make it practical for running in day-to-day development and CI (Travis). It leverages on `skaffold` to run the test suite from within the Kubernetes cluster itself.
The whole process, similar to the one used by [Sonobuoy](https://github.com/heptio/sonobuoy) when running the Kubernetes conformance test suite, is better understood by reading the included development documentation:

---

(...) to run the test suite against the resulting Minikube cluster, you may simply run:

```console
$ make e2e
```

Executing this command will do several things in background:

1. Create TLS secrets required for testing if they don't exist already.
1. Build the `nats-operator` binary locally.
1. Build a container image containing the `nats-operator` binary and deploy it to the Kubernetes cluster.
1. Build the `nats-operator-e2e` binary (which contains the test suite) locally.
1. Build a container image containing the `nats-operator-e2e` binary and deploy it to the Kubernetes cluster.
1. Wait for the `nats-operator` and `nats-operator-e2e` pods to be running.
1. Start streaming logs from the `nats-operator-e2e` pod and wait for it to terminate.
1. Delete the `nats-operator` and `nats-operator-e2e` pods.
1. Exit with the same exit code as the main container of the `nats-operator-e2e` pod.

---

Having this process in place, there is no reason to keep tests in `test/operator` outside the e2e suite as it only creates maintenance burden. Hence, those are being "ported" (re-written) here to `test/e2e`.

The TLS test that was previously skipped has also been ported and is now enabled again.

Most of the code was re-written to leverage on `ListWatcher`s instead of polling the Kubernetes API, which allows for cleaner, more readable tests.

Also, since the test suite now runs from within Kubernetes, usage of the `nats-ops` pod has been replaced with the official NATS Go client, which also adds up to making the tests more understandable and concise.

### CODE REVIEW

**Comment by pires** [comment]
Here's the output of a local run: ```
$ make e2e
Generating deepcopy funcs
Generating clientset for nats:v1alpha2 at github.com/nats-io/nats-operator/pkg/client/clientset
Generating listers for nats:v1alpha2 at github.com/nats-io/nats-operator/pkg/client/listers
Generating informers for nats:v1alpha2 at github.com/nats-io/nats-operator/pkg/client/informers
Starting build...
Found [minikube] context, using local docker daemon.
Building [gcr.io/project-id/nats-operator]...
Sending build context to Docker daemon  28.34MB
Step 1/6 : FROM gcr.io/distroless/base
latest: Pulling from distroless/base
8f125ded1b48: Pulling fs layer
8f125ded1b48: Verifying Checksum
8f125ded1b48: Download complete
8f125ded1b48: Pull complete
Digest: sha256:628939ac8bf3f49571d05c6c76b8688cb4a851af6c7088e599388259875bde20
Status: Downloaded newer image for gcr.io/distroless/base:latest
 ---> 1358ac6604c5
Step 2/6 : COPY build/nats-operator /nats-operator
 ---> 109f4c186bc8
Step 3/6 : ENV MY_POD_NAMESPACE default
 ---> Running in cca9ed1e23b0
 ---> ce254c31cf6d
Step 4/6 : ENV MY_POD_NAME nats-operator
 ---> Running in 669b78f44b98
 ---> 30c0ca56c6be
Step 5/6 : EXPOSE 8080
 ---> Running in a50644404a95
 ---> 45e1876afb40
Step 6/6 : CMD ["/nats-operator"]
 ---> Running in e8cc24c60f35
 ---> 3b71005fe887
Successfully built 3b71005fe887
Successfully tagged 5d663dddb27364fe8c2af7f0d3396143:latest
Build complete in 6.520433522s
Starting test...
Test complete in 7.594µs
Starting deploy...
kubectl client version: 1.10
kubectl version 1.12.0 or greater is recommended for use with skaffold
serviceaccount "nats-operator" created
clusterrolebinding.rbac.authorization.k8s.io "nats-operator-binding" created
clusterrole.rbac.authorization.k8s.io "nats-operator" created
pod "nats-operator" created
WARN[0010] error adding label to runtime object: patching resource /nats-operator-binding: the server could not find the requested resource
WARN[0011] error adding label to runtime object: patching resource /nats-operator: the server could not find the requested resource
Deploy complete in 3.245691305s
Starting build...
Found [minikube] context, using local docker daemon.
Building [gcr.io/project-id/nats-operator-e2e]...
Sending build context to Docker daemon  35.62MB
Step 1/3 : FROM gcr.io/distroless/base
 ---> 1358ac6604c5
Step 2/3 : COPY build/nats-operator-e2e /nats-operator-e2e
 ---> 826dd69463ca
Step 3/3 : CMD ["/nats-operator-e2e"]
 ---> Running in 0ea792ba1bb6
 ---> 71883f27e628
Successfully built 71883f27e628
Successfully tagged e95a2baad40ce02d62fe9a33527e5598:latest
Build complete in 1.451252142s
Starting test...
Test complete in 7.081µs
Starting deploy...
kubectl client version: 1.10
kubectl version 1.12.0 or greater is recommended for use with skaffold
pod "nats-operator-e2e" created
Deploy complete in 366.428583ms
=== RUN   TestCreateCluster
--- PASS: TestCreateCluster (27.65s)
=== RUN   TestPauseControl
--- PASS: TestPauseControl (138.02s)
=== RUN   TestConfigReloadOnResize
--- PASS: TestConfigReloadOnResize (19.91s)
=== RUN   TestConfigReloadOnClientAuthSecretChange
--- PASS: TestConfigReloadOnClientAuthSecretChange (68.80s)
=== RUN   TestConfigReloadOnNatsServiceRoleUpdates
--- PASS: TestConfigReloadOnNatsServiceRoleUpdates (70.10s)
=== RUN   TestConfigContainsFullMesh
--- PASS: TestConfigContainsFullMesh (10.99s)
=== RUN   TestExistingSecretIsReplaced
--- PASS: TestExistingSecretIsReplaced (3.30s)
=== RUN   TestResizeClusterFrom3To5
--- PASS: TestResizeClusterFrom3To5 (22.29s)
=== RUN   TestResizeClusterFrom5To3
--- PASS: TestResizeClusterFrom5To3 (23.78s)
=== RUN   TestCreateClusterWithTLSConfig
--- PASS: TestCreateClusterWithTLSConfig (12.62s)
=== RUN   TestUpgradeCluster
--- PASS: TestUpgradeCluster (19.92s)
PASS
ok  	command-line-arguments	398.748s
```

File: docs/development/development.md [hunk 1]
```diff
@@ -90,3 +90,39 @@ $ make run PROFILE=gke
...
+`nats-operator` includes an end-to-end test suite that is used to validate the implementation.
+As the test suite depends on having certain, advanced features enabled on the target Kubernetes cluster, Minikube is currently the only supported environment for running the test suite.
+To launch a Minikube cluster suitable for running the end-to-end test suite, you may run:
+
+```console
+$ minikube start \
+    --extra-config=apiserver.service-account-signing-key-file=/var/lib/minikube/certs/apiserver.key \
+    --extra-config=apiserver.service-account-issuer=api \
+    --extra-config=apiserver.service-account-api-audiences=api \
+    --extra-config=apiserver.service-account-key-file=/var/lib/minikube/certs/sa.pub \
```
Comment (wallyqs) [inline docs/development/development.md]: so the path in minikube changed for these in the latest version, I see now

Comment (pires) [reply docs/development/development.md]: We also needed to update the Ubuntu version.
